cmd/ecrecover.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright © 2022 NAME HERE <EMAIL ADDRESS>
+
+*/
+package cmd
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"log"
+
+	"github.com/spf13/cobra"
+)
+
+// Returns a signature string
+func PersonalSign(message string, privateKey *ecdsa.PrivateKey) (string, common.Hash, error) {
+	fullMessage := BuildPersonalSignMessage(message)
+	hash := crypto.Keccak256Hash([]byte(fullMessage))
+	signatureBytes, err := crypto.Sign(hash.Bytes(), privateKey)
+	if err != nil {
+		return "", hash, err
+	}
+	signatureBytes[64] += 27
+	return hexutil.Encode(signatureBytes), hash, nil
+}
+
+func BuildPersonalSignMessage(message string) string {
+	//It's metamask personal sign format.
+	fullMessage := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(message), message)
+	return fullMessage
+}
+func RecoverAddress(message string, signature string) (string, error) {
+	decode, err := hexutil.Decode(signature)
+	decode[64] -= 27
+	hash := crypto.Keccak256Hash([]byte(BuildPersonalSignMessage(message)))
+	sigPublicKey, err := crypto.SigToPub(hash.Bytes(), decode)
+	if err != nil {
+		return "", err
+	}
+	address := crypto.PubkeyToAddress(*sigPublicKey).Hex()
+	return address, nil
+}
+
+// ecrecoverCmd represents the ecrecover command
+var ecrecoverCmd = &cobra.Command{
+	Use:        "recover signature message",
+	Short:      "Recover address",
+	Long:       `Recover address from message and its signature`,
+	Args:       cobra.ExactArgs(2),
+	ArgAliases: []string{"signature", "message"},
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println("ecrecover called", args)
```

Comment:
Can use logrus so we can have fine grained control over logging output

Reply:
Used